experimental/design-notes.txt

```diff
@@ -0,0 +1,112 @@
+Counter       Counter
+Gauge         UpDownCounter or ValueRecorder?
+Summary       ValueRecorder
+Histogram     ValueRecorder
+NA            *Observer
+
+4. Prometheus uses Observe() API where OT uses Record() API. Prometheus doesn't appear to have asynchronous/callback style APIs.
+
+5. Prometheus has histogram and summaries configured with measurement precision and intervals at the point they are defined. OT and MM configure this via a
+separate configuration API.
+
+6. Prometheus has various helper utilities like NewTimer() that that handles details of timing code and then inputs the timed value into Histogram. This
+appears to avoid needing to declare a dedicated Timer Meter as MM did.
+
+
+
+AppMetrics - https://www.app-metrics.io/
+
+1. Naming - What OT calls instruments, AppMetrics calls metrics
+2. AppMetrics has an IMetrics interface somewhat similar to MeterProvider or MetricsRegistry in other libraries, however the provider isn't responsible for directly
+creating the API objects that represent metrics. Instead Metric definitions are created independent from any builder (similar to a LabelSet) and then passed as an
+argument to a measurement recording function. For example:
+IMetrics metrics = ...
+var counter = new CounterOptions { Name = "my_counter" } // this defines the metric
+metrics.Measure.Counter.Increment(counter); // increments the value of the metric stored in this IMetrics
+
+3. Metrics are never bound. Tags can be provided at the time a measurement is captured and merged with tags set via global tags in the IMetrics interface
+
+4. The metrics types are Counter, Gauge, Histogram, Meter, Timer, and Apdex (this is the 3rd distinct definition of the term Meter I've seen so far)
+Mapping is roughly
+Counter     UpDownCounter
+Gauge       ValueObserver or UpDownCounterObserver?
+Histogram   ValueRecorder
+Meter       NA  (it is close to a Counter, but it displays two stats: aggregate count and rate)
+Timer       NA
+Apdex       NA
+NA          Counter
+NA          CounterObserver
+
+
+OpenCensus - https://opencensus.io/stats/
+
+1. Naming - Instruments in OT are approximately Measures in OpenCensus. However Measures only define the name and units of measure, not the APIs that capture data.
+2. Unlike most other libraries, OpenCensus doesn't use different instrument types to imply intended aggregation for the collected data. There are only two types of measure,
+Double and long. The typical distinction between counters, gauges, histograms, etc are defined separately as aggregations using the View API.
+3. Measures are created directly from a static factory method and they are not associated with any particular collection pipeline. A separate StatsRecorder object provides
+the data collection API.
+4. The StatsRecorder collection API builds up a list of measurements first, then commits it with a single API call. I didn't see any simpler option presented in the samples
```

Comment:
We likely need a BatchRecord type functionality.  To allow our user to inject data for the same "timeframe".  I don't know what other scenarios requires "atomicity" (outside of marking same timeframe) but would like to find out more from others what the need here is.